### PR TITLE
Adding RNN encoder for LSTM-Transducer and LSTM-CTC models

### PR DESCRIPTION
**Pull request by VahidooX**
# What does this PR do ?

This PR adds RNN-based encoder to NeMo. It enables us to have LSTM-Transducer (RNN-T) and LSTM-CTC models. 

# Changelog 
- Added RNN encoder for ASR models (LSTM-T and LSTM-CTC models)
- Added stacking downsampling
- Added support for proj_size to Transducer decoders
- Added skip_nan_grad support for ASR models which skips the gradients when there is a nan or inf in the gradients

# Usage

```python
# Add a code snippet demonstrating how to use this 
```

**PR Type**:
- [x ] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: docs/source/asr/configs.rst [hunk 1]
```diff
@@ -511,13 +511,22 @@ Conformer-Transducer
 
 Please refer to the model page of `Conformer-Transducer <./models.html#Conformer-Transducer>`__ for more information on this model.
 
+RNN-Transducer
+~~~~~~~~~~~~~~
+
+The config files for RNN-Transducer (RNNT) and RNN-CTC models can be found at ``<NeMo_git_root>/examples/asr/conf/rnn/rnn_transducer_bpe.yaml`` and ``<NeMo_git_root>/examples/asr/conf/rnn/rnn_ctc_bpe.yaml`` respectively.
```
Comment (titu1994) [inline docs/source/asr/configs.rst:517]: Hmm considering our over use of RNN transducer, can we rename this model to (LN)LSTM Transducer? Highly confusing that we use RNN-T, RNNT, Transducer etc to refer to Conformer ContextNet and now LSTM encoders.

Comment (titu1994) [reply docs/source/asr/configs.rst:517]: Strictly speaking, its not the basic RNN encoder but an Layer Norm LSTM encoder with a transducer decoder.

File: docs/source/asr/models.rst [hunk 2]
```diff
@@ -127,6 +127,22 @@ You may find the example config files of Conformer-Transducer model with charact
 ``<NeMo_git_root>/examples/asr/conf/conformer/conformer_transducer_char.yaml`` and
 with sub-word encoding at ``<NeMo_git_root>/examples/asr/conf/conformer/conformer_transducer_bpe.yaml``.
 
+RNN-Transducer (RNNT)
```
Comment (titu1994) [inline docs/source/asr/models.rst:130]: Rename to LSTM Transducer?

File: docs/source/asr/models.rst [hunk 3]
```diff
@@ -127,6 +127,22 @@ You may find the example config files of Conformer-Transducer model with charact
...
+
+RNNT-Transducer (RNNT) is a model which uses RNNs (eg. LSTM) in the encoder. The architecture of this model is followed from suggestions in :cite:`asr-models-he2019streaming`.
+It uses RNNT/Transducer loss/decoder. The encoder consists of RNN layers (LSTM as default) with lower projection size to increase the efficiency.
+Layer norm is added between the layers to stabilize the training.
+It can be trained/used in unidirectional or bidirectional mode. The unidirectional mode is fully causal and can be used easily for simple and efficient frame-wise streaming. However the accuracy of this model is generally lower than other models like Conformer and Citrinet.
+
+This model supports both the sub-word level and character level encodings. You may find the example config file of RNNT model with wordpiece encoding at ``<NeMo_git_root>/examples/asr/conf/rnn_transducer_bpe.yaml``.
+You can find more details on the config files for the RNNT models at ``RNN-Transducer <./configs.html#rnn-transducer>``.
+
+RNN-CTC
```
Comment (titu1994) [inline docs/source/asr/models.rst:141]: Rename maybe?

File: examples/asr/conf/conformer/conformer_ctc_bpe.yaml [hunk 4]
```diff
@@ -30,6 +30,7 @@ model:
   sample_rate: 16000
   log_prediction: true # enables logging sample predictions in the output during training
   ctc_reduction: 'mean_batch'
+  skip_nan_grad: false
```
Comment (titu1994) [inline examples/asr/conf/conformer/conformer_ctc_bpe.yaml:33]: Any reason this is false? Also, I believe Megatron already has code to detect and skip batches with nan grads (and also correct the step counter for the scheduler)- Look at https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L342-L378 

Can we utilize similar technique at the ASRModel level so that automatically all models will benefit from this?

Comment (titu1994) [reply examples/asr/conf/conformer/conformer_ctc_bpe.yaml:33]: Also why is the default false? It should be good to always skip nan grads no?

File: examples/asr/conf/rnn/rnn_ctc_bpe.yaml [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    time_masks: 10 # set to zero to disable it
+    freq_width: 27
+    time_width: 0.05
+
+  encoder:
+    _target_: nemo.collections.asr.modules.RNNEncoder
+    feat_in: ${model.preprocessor.features}
+    n_layers: 8
+    d_model: 2048
+    proj_size: 640 # you may set it if you need different output size other than the default d_model
```
Comment (titu1994) [inline examples/asr/conf/rnn/rnn_ctc_bpe.yaml:88]: Is this projection only for the last layer or the intermediate layers as well similar to TF projection LSTMs?

File: examples/asr/conf/rnn/rnn_transducer_bpe.yaml [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    freq_masks: 2 # set to zero to disable it
+    time_masks: 10 # set to zero to disable it
+    freq_width: 27
+    time_width: 0.05
+
+  encoder:
+    _target_: nemo.collections.asr.modules.RNNEncoder
+    feat_in: ${model.preprocessor.features}
+    n_layers: 8
+    d_model: 2048
```
Comment (titu1994) [inline examples/asr/conf/rnn/rnn_transducer_bpe.yaml:92]: Use ${model.model_defaults.xyz} here

File: examples/asr/conf/rnn/rnn_transducer_bpe.yaml [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+      joint_hidden: ${model.model_defaults.joint_hidden}
+      activation: "relu"
+      dropout: 0.2
+
+  decoding:
+    strategy: "greedy_batch" # can be greedy, greedy_batch, beam, tsd, alsd.
+
+    # greedy strategy config
+    greedy:
+      max_symbols: 30
```
Comment (titu1994) [inline examples/asr/conf/rnn/rnn_transducer_bpe.yaml:144]: Set to 5 ot 10, no need for such large value actually. We can change for older configs later.

File: nemo/collections/asr/models/rnnt_bpe_models.py [hunk 8]
```diff
@@ -179,7 +179,7 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             cfg.joint.num_classes = len(vocabulary)
             cfg.joint.vocabulary = ListConfig(list(vocabulary))
             cfg.joint.jointnet.encoder_hidden = cfg.model_defaults.enc_hidden
-            cfg.joint.jointnet.pred_hidden = cfg.model_defaults.pred_hidden
+            cfg.joint.jointnet.pred_hidden = cfg.decoder.prednet.get('proj_size', cfg.model_defaults.pred_hidden)
```
Comment (titu1994) [inline nemo/collections/asr/models/rnnt_bpe_models.py:182]: Revert. We use model_defaults for exactly this reason. Inject values from model_defaults into other parts of the config as needed. The values of `pred_hidden` or `enc_hidden` or `joint_hidden` are top level values and must be used as is, common to all RNNT models.

```yaml

  model_defaults:
    enc_hidden: ${model.encoder.d_model}
    pred_hidden: 640
    joint_hidden: 640

```

File: nemo/collections/asr/models/rnnt_models.py [hunk 9]
```diff
@@ -74,7 +74,9 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             self.cfg.joint.num_classes = len(self.cfg.labels)
             self.cfg.joint.vocabulary = self.cfg.labels
             self.cfg.joint.jointnet.encoder_hidden = self.cfg.model_defaults.enc_hidden
-            self.cfg.joint.jointnet.pred_hidden = self.cfg.model_defaults.pred_hidden
+            self.cfg.joint.jointnet.pred_hidden = self.cfg.decoder.prednet.get(
```
Comment (titu1994) [inline nemo/collections/asr/models/rnnt_models.py:77]: Revert, explained above.

File: nemo/collections/asr/modules/rnn_encoder.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+    def __init__(
+        self,
+        feat_in,
+        n_layers,
+        d_model,
+        proj_size=-1,
+        rnn_type='lstm',
+        bidirectional=True,
+        subsampling='striding',
```
Comment (titu1994) [inline nemo/collections/asr/modules/rnn_encoder.py:100]: Add types to these arguments

File: nemo/collections/asr/modules/rnn_encoder.py [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        subsampling_factor=4,
+        subsampling_conv_channels=-1,
+        dropout=0.1,
```
Comment (titu1994) [inline nemo/collections/asr/modules/rnn_encoder.py:103]: Defaults in docstring differ from code.

File: nemo/collections/asr/modules/rnn_encoder.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+                    subsampling_factor=subsampling_factor, feat_in=feat_in, feat_out=proj_size
+                )
+            else:
+                self.pre_encode = ConvSubsampling(
+                    subsampling=subsampling,
+                    subsampling_factor=subsampling_factor,
+                    feat_in=feat_in,
+                    feat_out=proj_size,
+                    conv_channels=subsampling_conv_channels,
+                    activation=nn.ReLU(),
```
Comment (titu1994) [inline nemo/collections/asr/modules/rnn_encoder.py:124]: Not swish? Should this be user-modifiable ?

File: nemo/collections/asr/modules/rnn_encoder.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+                    batch_first=True,
+                    bidirectional=bidirectional,
+                    proj_size=rnn_proj_size,
+                )
+            self.layers.append(layer)
+            self.layers.append(nn.LayerNorm(proj_size))
+            self.layers.append(nn.Dropout(p=dropout))
+            self._feat_out = proj_size
+
+        if proj_size > 0 and self._feat_out != proj_size:
```
Comment (titu1994) [inline nemo/collections/asr/modules/rnn_encoder.py:155]: Ah ok final projection only, hmm then just overload the `pred_hidden` of `model_defaults` here.

File: nemo/collections/asr/modules/rnn_encoder.py [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+
+        audio_signal = torch.transpose(audio_signal, 1, 2)
+
+        if isinstance(self.pre_encode, ConvSubsampling) or isinstance(self.pre_encode, StackingSubsampling):
+            audio_signal, length = self.pre_encode(audio_signal, length)
+        else:
+            audio_signal = self.pre_encode(audio_signal)
+
+        for lth, layer in enumerate(self.layers):
+            if isinstance(layer, torch.nn.Dropout) or isinstance(layer, torch.nn.LayerNorm):
```
Comment (titu1994) [inline nemo/collections/asr/modules/rnn_encoder.py:178]: For cleanliness, how about doing outputs = layer(audio_signal) and then check if type(outputs) in (list, tuple), then unpack the 0th element for LSTM/GRU/RNN etc?

File: nemo/core/classes/modelPT.py [hunk 15]
```diff
@@ -1342,3 +1342,17 @@ def update_save_restore_connector(cls, save_restore_connector):
             cls._save_restore_connector = save_restore_connector
         else:
             setattr(cls, '_save_restore_connector', save_restore_connector)
+
+    def on_after_backward(self):
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:1346]: Dont do this at ModelPT level, Megatron has different logic to handle this but calls super() which will inturn call this and cause issues. Push it up to ASRModel. Will need more discussion from team to put it into ModelPT.

**Review by titu1994 — COMMENTED** [review]
Overall looks great, minor comments here and there.

File: examples/asr/conf/rnn/rnn_ctc_bpe.yaml (hunk 5 above)
Comment (VahidooX) [reply examples/asr/conf/rnn/rnn_ctc_bpe.yaml:88]: This is after every LSTM layers to reduce computation cost. That is the same for both encoder and decoder.

File: nemo/collections/asr/modules/rnn_encoder.py (hunk 11 above)
Comment (VahidooX) [reply nemo/collections/asr/modules/rnn_encoder.py:103]: fixed.

File: nemo/collections/asr/modules/rnn_encoder.py (hunk 14 above)
Comment (VahidooX) [reply nemo/collections/asr/modules/rnn_encoder.py:178]: Good idea! changed it.

File: nemo/core/classes/modelPT.py (hunk 15 above)
Comment (VahidooX) [reply nemo/core/classes/modelPT.py:1346]: For now, I dropped it as it was not necessary for this model. It did not help this model. We can discuss this and add it in a separate PR.

Comment (VahidooX) [reply nemo/core/classes/modelPT.py:1346]: Moved it to ASRModel for now.

File: examples/asr/conf/conformer/conformer_ctc_bpe.yaml (hunk 4 above)
Comment (VahidooX) [reply examples/asr/conf/conformer/conformer_ctc_bpe.yaml:33]: It makes training slightly slower and it did not help this model.

File: docs/source/asr/models.rst (hunk 3 above)
Comment (VahidooX) [reply docs/source/asr/models.rst:141]: changed names.

File: docs/source/asr/configs.rst (hunk 1 above)
Comment (VahidooX) [reply docs/source/asr/configs.rst:517]: changed names.

File: nemo/collections/asr/models/rnnt_bpe_models.py (hunk 8 above)
Comment (VahidooX) [reply nemo/collections/asr/models/rnnt_bpe_models.py:182]: reverted.

File: nemo/collections/asr/modules/rnn_encoder.py (hunk 10 above)
Comment (VahidooX) [reply nemo/collections/asr/modules/rnn_encoder.py:100]: added.

File: examples/asr/conf/rnn/rnn_transducer_bpe.yaml (hunk 6 above)
Comment (VahidooX) [reply examples/asr/conf/rnn/rnn_transducer_bpe.yaml:92]: I mean in the ctc configs. For transducer configs we have them as we need them for decoder parts.

File: nemo/collections/asr/models/rnnt_bpe_models.py [hunk 16]
```diff
@@ -193,7 +193,7 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             cfg.joint.num_classes = len(vocabulary)
             cfg.joint.vocabulary = ListConfig(list(vocabulary))
             cfg.joint.jointnet.encoder_hidden = cfg.model_defaults.enc_hidden
-            cfg.joint.jointnet.pred_hidden = cfg.model_defaults.pred_hidden
+            cfg.joint.jointnet.pred_hidden = cfg.model_defaults.get('proj_size', cfg.model_defaults.pred_hidden)
```
Comment (titu1994) [inline nemo/collections/asr/models/rnnt_bpe_models.py:196]: pred_hidden is necessary for all rnnt configs, why not just use that instead of yet another variable called proj_size?

File: nemo/collections/asr/models/rnnt_models.py [hunk 17]
```diff
@@ -74,7 +74,9 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             self.cfg.joint.num_classes = len(self.cfg.labels)
             self.cfg.joint.vocabulary = self.cfg.labels
             self.cfg.joint.jointnet.encoder_hidden = self.cfg.model_defaults.enc_hidden
-            self.cfg.joint.jointnet.pred_hidden = self.cfg.model_defaults.pred_hidden
+            self.cfg.joint.jointnet.pred_hidden = self.cfg.model_defaults.get(
+                'proj_size', self.cfg.model_defaults.pred_hidden
```
Comment (titu1994) [inline nemo/collections/asr/models/rnnt_models.py:78]: Same as above

File: nemo/collections/asr/modules/rnn_encoder.py (hunk 12 above)
Comment (titu1994) [reply nemo/collections/asr/modules/rnn_encoder.py:122]: Lets experiment first, we can make it user configurable via config later if it helps.

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall looks good, just reuse `model_defaults.pred_hidden` instead of add yet another variable `proj_size` at model_defaults level.

Things in model_defaults are : 
1) Global to a model archetype (not just 1 model like LSTMs here)
2) Referred to multiple places in the model config (so user can modify the global one and propagate to rest of config).

proj_size is effectively pred_hidden here so I don't see why should split them into two.

**Review by titu1994 — APPROVED** [review]
Looks great ! Merge when ready

**Review by titu1994 — APPROVED** [review]
